calico_containers/tests/unit/calicoctl_test.py

```diff
@@ -176,6 +181,708 @@ def test_validate_arguments(self, case, sys_exit_called):
             # Assert method exits if bad input
             self.assertEqual(m_sys_exit.called, sys_exit_called)
 
+    @patch('calico_ctl.container.enforce_root', autospec=True)
+    @patch('calico_ctl.container.get_container_info_or_exit', autospec=True)
+    @patch('calico_ctl.container.client', autospec=True)
+    @patch('calico_ctl.container.get_pool_or_exit', autospec=True)
+    @patch('calico_ctl.container.netns', autospec=True)
+    def test_container_add(self, m_netns, m_get_pool_or_exit, m_client,
+                           m_get_container_info_or_exit, m_enforce_root):
+        # Set up mock objects
+        m_get_container_info_or_exit.return_value = {
+            'Id': 666,
+            'State': {'Running': 1, 'Pid': 'Pid_info'}
+        }
+        m_client.get_endpoint.side_effect = KeyError
+        m_client.get_default_next_hops.return_value = 'next_hops'
+        m_netns.set_up_endpoint.return_value = 'endpoint'
+
+        # Call method under test
+        test_return = container.container_add('container1', '1.1.1.1', 'interface')
+
+        # Assert
+        m_enforce_root.assert_called_once_with()
+        m_get_container_info_or_exit.assert_called_once_with('container1')
+        m_client.get_endpoint.assert_called_once_with(
+            hostname=node.hostname,
+            orchestrator_id=utils.ORCHESTRATOR_ID,
+            workload_id=666
+        )
+        m_get_pool_or_exit.assert_called_once_with(IPAddress('1.1.1.1'))
+        m_client.get_default_next_hops(node.hostname)
+        m_netns.set_up_endpoint(
+            ip='1.1.1.1',
+            hostname=node.hostname,
+            orchestrator_id=utils.ORCHESTRATOR_ID,
+            workload_id=666,
+            cpid='Pid_info',
+            next_hop_ips='next_hops',
+            veth_name='interface',
+            proc_alias='/proc/'
+        )
+        m_client.set_endpoint.assert_called_once_with('endpoint')
+        self.assertEqual(test_return, 'endpoint')
+
+    @patch('sys.exit', autospec=True)
+    @patch('calico_ctl.container.enforce_root', autospec=True)
+    @patch('calico_ctl.container.get_container_info_or_exit', autospec=True)
+    @patch('calico_ctl.container.client', autospec=True)
+    @patch('calico_ctl.container.get_pool_or_exit', autospec=True)
+    @patch('calico_ctl.container.netns', autospec=True)
```

Comment:
I think you should try to have as few patches as possible.  Since this case only goes to the first sys.exit, you can remove some of these patches because they won't be hit.  You'll have to change the assert below too...